src/main/java/com/arangodb/springframework/core/util/ArangoExceptionTranslator.java

```diff
@@ -20,63 +20,76 @@
 
 /**
+ * Translate any {@link ArangoDBException} to the appropriate {@link DataAccessException} using response code and error number.
+ *
  * @author Mark Vollmary
  * @author Christian Lechner
- *
+ * @author Arne Burmeister
  */
 public class ArangoExceptionTranslator implements PersistenceExceptionTranslator {
 
 	@Override
+	@Nullable
 	public DataAccessException translateExceptionIfPossible(final RuntimeException ex) {
-		DataAccessException dae = null;
-		if (ex instanceof DataAccessException) {
-			dae = DataAccessException.class.cast(ex);
-		} else if (ex instanceof ArangoDBException) {
-			final ArangoDBException exception = ArangoDBException.class.cast(ex);
+		if (ex instanceof DataAccessException exception) {
+			return exception;
+		}
+		if (ex instanceof ArangoDBException exception) {
 			final Integer responseCode = exception.getResponseCode();
-			if (responseCode != null) {
-				switch (responseCode) {
-				case ArangoErrors.ERROR_HTTP_UNAUTHORIZED:
-				case ArangoErrors.ERROR_HTTP_FORBIDDEN:
-					dae = new PermissionDeniedDataAccessException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_BAD_PARAMETER:
-				case ArangoErrors.ERROR_HTTP_METHOD_NOT_ALLOWED:
-					dae = new InvalidDataAccessApiUsageException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_NOT_FOUND:
-					dae = new InvalidDataAccessResourceUsageException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_CONFLICT:
-					dae = new DataIntegrityViolationException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_PRECONDITION_FAILED:
-				case ArangoErrors.ERROR_HTTP_SERVICE_UNAVAILABLE:
-					dae = new DataAccessResourceFailureException(exception.getMessage(), exception);
-					break;
-				case ArangoErrors.ERROR_HTTP_SERVER_ERROR:
-				default:
-					dae = new ArangoUncategorizedException(exception.getMessage(), exception);
-					break;
-				}
+			if (responseCode == null) {
+				return new ArangoUncategorizedException(exception.getMessage(), exception);
 			}
+			BiFunction<String, ArangoDBException, DataAccessException> constructor = switch (responseCode) {
+				case ERROR_HTTP_UNAUTHORIZED, ERROR_HTTP_FORBIDDEN -> PermissionDeniedDataAccessException::new;
+				case ERROR_HTTP_BAD_PARAMETER, ERROR_HTTP_METHOD_NOT_ALLOWED -> InvalidDataAccessApiUsageException::new;
+				case ERROR_HTTP_NOT_FOUND -> mostSpecific(exception, Map.ofEntries(
+							entry(hasErrorNumber(ERROR_ARANGO_DOCUMENT_NOT_FOUND), DataRetrievalFailureException::new)
+						), InvalidDataAccessResourceUsageException::new);
+				case ERROR_HTTP_CONFLICT -> mostSpecific(exception, Map.ofEntries(
+							entry(hasErrorNumber(ERROR_ARANGO_CONFLICT).and(errorMessageContains("write-write")), TransientDataAccessResourceException::new),
+							entry(hasErrorNumber(ERROR_ARANGO_CONFLICT).and(errorMessageContains("_rev")), OptimisticLockingFailureException::new),

```

Comment:
```suggestion
```

Is this needed? In which case ArangoDB returns `code: 409, errorNum: 1200` for mismatching `_rev`?
AFAICS updating a document with a mismatching `_rev` returns: `code: 412, errorNum: 1200, errorMessage: conflict, _rev values do not match`.

Reply:
Yes, we had this case when using `save()` aka `REPSERT` and passing `_rev`